build-logic/integration-testing/src/main/kotlin/gradlebuild/integrationtests/tasks/EmitTeamCityImportDataServiceMessageTask.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gradlebuild.integrationtests.tasks
+
+import gradlebuild.basics.BuildEnvironment
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
+
+/**
+ * This is a workaround for https://youtrack.jetbrains.com/issue/TW-76894.
+ *
+ * In short, we want TeamCity to be aware of the test execution data (which tests are executed and how long they are),
+ * even when the Test task is `FROM-CACHE` or `UP-TO-DATE`. This task is registered as a finalizer for the test task
+ * and will output a service message to instruct TeamCity to read JUnit test result XMLs.
+ *
+ * See https://www.jetbrains.com/help/teamcity/service-messages.html#Importing+XML+Reports
+ */
+@DisableCachingByDefault(because = "It outputs a TeamCity service message")
+abstract class EmitTeamCityImportDataServiceMessageTask : DefaultTask() {
```

Comment:
We talked about using `org.gradle.build.event.BuildEventsListenerRegistry#onTaskCompletion` for this instead of a finalizer task.